File: python3/02_semantic/machine.py
from typing import Dict
from statement import StmtMixin


class Machine:
    def __init__(self, stmt: StmtMixin, env: Dict):
        self.stmt = stmt
        self.env = env

    def step(self):
        self.stmt, self.env = self.stmt.reduce(self.env)

    def run(self, debug=True):
        while self.stmt.reducible:
            if debug:
                print(self.stmt, self.env)
            self.step()

        if debug:
            print(self.stmt, self.env)

File: computation-go/semantic/small_step_semantic/machine.go
package small_step_semantic

import "fmt"

type Machine struct {
	stmt Stmter
	env  map[string]Exprer
}

func NewMachine(stmt Stmter, env map[string]Exprer) *Machine {
	return &Machine{stmt: stmt, env: env}
}

func (m *Machine) Step() {
	m.stmt, m.env = m.stmt.Reduce(m.env)
}

func (m *Machine) Run() {
	for m.stmt.Reducible() {
		fmt.Println(m.stmt, m.env)
		m.Step()
	}
	fmt.Println(m.stmt, m.env)
}
